### DISCUSSION TITLE
Linearize do-mpc ODE and measurement equations

### DISCUSSION THREAD

**Discussion by 4flixt** [do-mpc/do-mpc#278] [Show and tell]
We have been asked many times how to obtain a linearized system from a do-mpc model. While we are working on a more convenient feature, there is a nice workaround that might come in handy for some applications.

Assume we have created and setup a model instance named ``model``.

```python
# Get all variable(types) known to the model
x, u, z, tvp, p, w = model['x', 'u', 'z', 'tvp', 'p', 'w']
# Get the RHS of the ODE and the (nonlinear) measurement equations
f = model._rhs
h = model._y_expression
# Create symbolic expression for the derivative of the RHS w.r.t to the states, inputs
df_dx = jacobian(f, x)
df_du = jacobian(f, u)
# Create symbolic expression for the derivative of the measurement equations w.r.t to the states, inputs
dh_dx = jacobian(h, x)
dh_du = jacobian(h, u)

# Create function that return the derivative of the RHS and the meas expression for a given state, input, ...
df_dx_fun = Function('df_dx_fun', [x, u, z, tvp, p, w], [df_dx])
df_du_fun = Function('df_du_fun', [x, u, z, tvp, p, w], [df_du])
dh_dx_fun = Function('dh_dx_fun', [x, u, z, tvp, p, w], [dh_dx])
dh_du_fun = Function('dh_du_fun', [x, u, z, tvp, p, w], [dh_du])
# Create concrete values for states, inputs etc. ---> You want to change these later, depending where you want to linearize
x0 = x(0)
u0 = u(0)
z0 = z(0)
tvp0 = tvp(0)
p0 = p(0)
w0 = w(0)

# Create matrices A,B,C,D for the linearized system
A = df_dx_fun(x0, u0, z0, tvp0, p0, w0)
B = df_du_fun(x0, u0, z0, tvp0, p0, w0)
C = dh_dx_fun(x0, u0, z0, tvp0, p0, w0)
D = dh_du_fun(x0, u0, z0, tvp0, p0, w0)

# Show the linearized system
print('A: ', A)
print('B: ', B)
print('C: ', C)
print('D: ', D)
```

Notice that ``x(0)`` creates a numeric structure that can be indexed similarly as shown [here](https://www.do-mpc.com/en/latest/api/do_mpc.model.Model.x.html). By passing the ``0``, all values are set to zero. For a concrete application you could:
- Create ``x0=x(0)`` and set values on ``x0`` with proper indexing, e.g. ``x0['name_of_state_1'] = 5``
- Have a vector of states with numeric values, e.g. ``x_val = np.array([1,2,3,4])`` and obtain ``x0 = x(x_val)``

**Comment by jrbull** [comment]
Thank you, this is exactly what I needed to make an EKF for a DAE model.

**Reply by jrbull** [reply]
When I tested this exact code on my model, df_dx_fun(x0, u0, z0, tvp0, p0, w0) evaluated to numbers, but dh_dx_fun(x0, u0, z0, tvp0, p0, w0) evaluated to 0s and 1s. To get numerical values for the measurements, I had to use h = model._alg and select components that corresponded to the measurement variables - this, however, without the added noise. Have I done something wrong?

**Comment by Mrfeli01** [comment]
Hi Jonathan, 

this is quite an old post. The newest version of do-mpc has pre implemented functions for [linerization ](https://www.do-mpc.com/en/latest/api/do_mpc.model.linearize.html) as well as an [EKF](https://www.do-mpc.com/en/latest/api/do_mpc.estimator.EKF.html). However, our [EKF](https://www.do-mpc.com/en/latest/api/do_mpc.estimator.EKF.html)  implementastion is not working for dae systems. Can you reduce it to an ode? There is a do-mpc function for this [too](https://www.do-mpc.com/en/latest/api/do_mpc.model.dae2odeconversion.html).

Maybe it would be easier to discuss your problem in a short call to see if I can help:
felix.brabender@tu-dortmund.de

Best
Felix
